rfcs/records.md

```diff
@@ -0,0 +1,196 @@
+
+Reading and writing fields from a record uses `.` or `[]` operator; like tables, passing the field name returns the field value. Unlike tables, if the field is not
+present in the table, the error is raised. This is in constrast with tables where `nil` is returned for unknown keys upon read; records are meant to be stricted than
+tables and as such returning `nil` will mask valuable errors, and make it more difficult to be strict about the types of the result.
+
+Invoking methods with `:` desugars into `getmetatable(obj).__index.method(obj, args)` instead of the usual `obj.method(obj, args)`. This is important because it allows to
+keep the method calls as efficient as possible, as they don't need to check whether the object has a given method.
+
+> TODO: Should we use `__namecall` instead of `__index`? It seems more consistent, but at the same time `__namecall` today expects a function so it might be best
+> to leave it as is?
+
+> TODO: How do we expose the record keys? Should it be a builtin? Accessible through shape? Not available initially?
+
+### Defining and constructing records
+
+To define a record, you need to create the shape, which you can do using the newly introduced syntax with a context-sensitive `record` keyword:
+
+Syntax A:
+
+```
+record Person = { name: string, age: number }
+
+-- types can be omitted
+record Point = { x, y }
+```
+
+Syntax B:
+
+```
+record Person(name: string, age: number)
+-- types can be omitted
+record Point(x, y)
+```
+
+This defines `Point` simultaneously as a local variable that corresponds to the shape table, and a type variable that corresponds to the record type.
+
+The resulting shape table automatically is set up to be a valid record shape, but can still be modified by adding methods to it:
+
+```
+function Point.__add(l, r)
+    return Point(l.x + r.x, l.y + r.y)
+end
+
+function Point:sum()
+    return self.x + self.y
+end
+
+function Point.newDiagonal(v)
+    return Point(v, v)
+end
+```
+
+Note that `Point` is simply a table and as such it can be used to store static methods as well; as it also serves as a metatable, metafields defined on this table
+will change the behavior of the record values. The shape isn't frozen automatically but can be frozen manually if desired via `table.freeze`.
+
+To create a record, you need to use a record constructor. This is where the draft design has four options, each goes with one of the syntax options.
+
+A1. Creation uses special syntax, `new Record { field = value ... }`. This is unsurprising and easy to implement, but verbose.
+A2. Creation uses existing Lua DSL syntax `Record { field = value, ... }`. This is concise but requires a slightly intricate bytecode design to keep efficient.
+B1. Creation uses special syntax, `new Record(value, ...)`. This is unsurprising and easy to implement, but verbose.
+B2. Creation uses existing Lua call syntax, `Record(value, ...)`. This is concise and reasonably easy to keep efficient.
```

Comment:
Could we implement option A2 as syntax sugar for option B2?

Also, if we pursued B2, would `Record` still act as a normal function value at runtime?  eg

```lua
function apply(f, x)
    return f(x)
end

record Seconds { m: number }

local d = apply(Seconds, 4)
```

Reply:
I think the implementation detail for A2 and B2 would be that the record shape would be callable via `__call`, so it could be used as a first class callable object if need be, although I'd need to think through some details on this.

Reply:
I find B2 to be really compelling for avoiding complexity, even though I prefer explicitly specifying the keys. If a developer wants to add a function call constructor, they could always bang out a method like:
```lua
function Point.new(x, y): Point
    return new Point { x = x, y = y }
end
```
which is similar to the current state today.

Reply:
Can you clarify which syntactic variant you prefer? Asking because you prefer B2 but .new is using A1

Reply:
Resolving because I'm removing the `new` operator and just leaving the options A & B, feel free to comment on the updated version.